File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys
from collections import deque


def bfs(y, x, height):
    global n, buildings, visit
    que = deque()
    que.append([y, x])
    dy, dx = [0, 0, -1, 1], [1, -1, 0, 0]
    visit[y][x] = True
    while que:
        cur_y, cur_x = que.popleft()
        for i in range(4):
            next_y, next_x = dy[i] + cur_y, dx[i] + cur_x
            if 0 <= next_y < n and 0 <= next_x < n \
                    and not visit[next_y][next_x] and height < buildings[next_y][next_x]:
                visit[next_y][next_x] = True
                que.append([next_y, next_x])


def counting_area(height):
    global buildings, n, visit, answer
    cnt = 0
    for i in range(n):
        for j in range(n):
            if not visit[i][j] and buildings[i][j] > height:
                bfs(i, j, height)
                cnt += 1
    answer = answer if answer > cnt else cnt


n = int(sys.stdin.readline())
buildings = []
max_height = 0
answer = 0

for _ in range(n):
    input_list = list(map(int, sys.stdin.readline().split()))
    comp_height = max(input_list)
    max_height = max_height if max_height > comp_height else comp_height
    buildings.append(input_list)

for h in range(max_height):
    visit = [[False] * n for _ in range(n)]
    counting_area(h)

print(answer)

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

class Main{
  private static int n;
  private static int[][] buildings;
  private static boolean[][] visit;
  public static void main(String[] args) throws IOException{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    n = Integer.parseInt(br.readLine());
    buildings = new int[n][n];
    int maxHeight = 0;
    for(int i=0 ; i<n ; i++){
      String[] values = br.readLine().split(" ");
      for(int j=0 ; j<n ; j++){
        buildings[i][j] = Integer.parseInt(values[j]);
        maxHeight = Math.max(maxHeight, buildings[i][j]);
      }
    }

    int answer = 0;
    for(int h=1 ; h<=maxHeight ; h++){
      visit = new boolean[n][n];
      int cnt = 0;
      for(int i=0 ; i<n ; i++){
        for(int j=0 ; j<n ; j++){
          if(buildings[i][j] < h || visit[i][j]) continue;
          cnt++;
          bfs(i, j, h);
        }
      }
      answer = Math.max(answer, cnt);
    }

    System.out.println(answer);
    br.close();
  }

  private static void bfs(int sy, int sx, int h){
    Queue<int[]> que = new LinkedList<>();
    que.add(new int[]{sy, sx});
    final int[] dy = {0,0,1,-1}, dx = {1,-1,0,0};
    visit[sy][sx] = true;

    while(!que.isEmpty()){
      int[] cur = que.poll();
      int y =  cur[0], x = cur[1];
      for(int i=0 ; i<4 ; i++){
        int ny = y + dy[i], nx = x + dx[i];
        if(ny < 0 || ny >= n || nx < 0 || nx >= n || visit[ny][nx] || h > buildings[ny][nx])
          continue;
        que.add(new int[]{ny, nx});
        visit[ny][nx] = true;
      }
    }
  }
}